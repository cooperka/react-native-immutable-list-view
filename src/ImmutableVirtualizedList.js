import Immutable from 'immutable';
import React, { PureComponent } from 'react';

// This file and its related dependencies needs to be downloaded into your app's `node_modules`.
// See the README for instructions.
//
// eslint-disable-next-line import/no-unresolved, import/extensions
import VirtualizedList from 'react-native/Libraries/Lists/VirtualizedList';

import utils from './utils';

/**
 * A VirtualizedList capable of displaying {@link https://facebook.github.io/immutable-js/ Immutable} data
 * out of the box.
 */
// eslint-disable-next-line react/prefer-stateless-function
class ImmutableVirtualizedList extends PureComponent {

  static propTypes = {
    // Pass through any props that VirtualizedList would normally take.
    ...VirtualizedList.propTypes,

    /**
     * The immutable data to be rendered in a VirtualizedList.
     */
    // eslint-disable-next-line consistent-return
    immutableData: (props, propName, componentName) => {
      // Note: It's not enough to simply validate PropTypes.instanceOf(Immutable.Iterable),
      // because different imports of Immutable.js across files have different class prototypes.
      // TODO: Add support for Immutable.Map, etc.
      if (Immutable.Map.isMap(props[propName])) {
        return new Error(`Invalid prop ${propName} supplied to ${componentName}: Support for Immutable.Map is coming soon. For now, try an Immutable List, Set, or Range.`);
      } else if (!utils.isImmutableIterable(props[propName])) {
        return new Error(`Invalid prop ${propName} supplied to ${componentName}: Must be instance of Immutable.Iterable.`);
      }
    },
  };

  /**
   * Provides a handle to the underlying scroll responder.
   * Note that `this.virtualizedListRef` might not be a `ScrollView`, so we
   * need to check that it responds to `getScrollResponder` before calling it.
   */
  getScrollResponder() {
    if (this.virtualizedListRef && this.virtualizedListRef.getScrollResponder) {
      return this.listViewRef.getScrollResponder();
    }
    return null;
  }

  getScrollableNode() {
    if (this.virtualizedListRef && this.virtualizedListRef.getScrollableNode) {
      return this.listViewRef.getScrollableNode();
    }
    return null;
  }

  scrollToEnd(...args) {
    return this.virtualizedListRef && this.virtualizedListRef.scrollToEnd(...args);
  }

  scrollToIndex(...args) {
    return this.virtualizedListRef && this.virtualizedListRef.scrollToIndex(...args);
  }

  scrollToItem(...args) {
    return this.virtualizedListRef && this.virtualizedListRef.scrollToItem(...args);
  }

  scrollToOffset(...args) {
    return this.virtualizedListRef && this.virtualizedListRef.scrollToOffset(...args);
  }

  recordInteraction(...args) {
    return this.virtualizedListRef && this.virtualizedListRef.recordInteraction(...args);
  }

  render() {
    const { immutableData } = this.props;

    return (
      <VirtualizedList
        ref={(virtualizedList) => { this.virtualizedListRef = virtualizedList; }}
        data={immutableData}
        getItem={(items, index) => utils.getValueFromKey(index, items)}
        getItemCount={(items) => (items.size || items.length || 0)}
        keyExtractor={(item, index) => String(index)}
        {...this.props}
      />
    );
  }

}

export default ImmutableVirtualizedList;
