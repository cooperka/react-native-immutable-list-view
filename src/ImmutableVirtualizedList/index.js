export { default } from './ImmutableVirtualizedList';
export * from './EmptyVirtualizedList';
