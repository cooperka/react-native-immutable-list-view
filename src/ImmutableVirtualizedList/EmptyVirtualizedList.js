import Immutable from 'immutable';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import { Text, VirtualizedList } from 'react-native';

import ImmutableVirtualizedList from './ImmutableVirtualizedList';

import styles from '../styles';
import utils from '../utils';

/**
 * A VirtualizedList that displays a single item showing that there is nothing to display.
 * Useful e.g. for preserving the ability to pull-refresh an empty list.
 */
class EmptyVirtualizedList extends PureComponent {

  static propTypes = {
    // Pass through any props that VirtualizedList would normally take.
    ...VirtualizedList.propTypes,

    // Make this prop optional instead of required.
    renderItem: PropTypes.func,

    emptyText: PropTypes.string,
  };

  static defaultProps = {
    ...VirtualizedList.defaultProps,

    keyExtractor: (_, index) => String(index),

    emptyText: 'No data.',
  };

  state = {
    listData: utils.UNITARY_LIST,
  };

  componentWillMount() {
    this.setListDataFromProps(this.props);
  }

  componentWillReceiveProps(nextProps) {
    this.setListDataFromProps(nextProps);
  }

  setListDataFromProps(props) {
    const { listData } = this.state;
    const { renderEmpty, renderEmptyInList, emptyText } = props;

    // Update the data to make sure the list re-renders if any of the relevant props have changed.
    this.setState({
      listData: listData.set(0, Immutable.fromJS([renderEmpty, renderEmptyInList, emptyText])),
    });
  }

  /**
   * Returns a simple text element showing the `emptyText` string.
   * This method can be overridden by passing in your own `renderItem` prop instead.
   */
  renderItem() {
    const { emptyText } = this.props;

    return (
      <Text style={styles.emptyText}>
        {emptyText}
      </Text>
    );
  }

  render() {
    const { listData } = this.state;
    const { renderEmpty, renderEmptyInList, ...passThroughProps } = this.props;

    return (
      <ImmutableVirtualizedList
        renderItem={() => this.renderItem()}
        {...passThroughProps}
        immutableData={listData}
      />
    );
  }

}

export { EmptyVirtualizedList };
