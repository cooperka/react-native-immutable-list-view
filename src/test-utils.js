/* eslint-disable import/no-extraneous-dependencies */
import Immutable from 'immutable';
import React from 'react';
import { Text, ListView } from 'react-native';
import renderer from 'react-test-renderer';
/* eslint-enable */

import ImmutableListView from './ImmutableListView';
import ImmutableVirtualizedList from './ImmutableVirtualizedList';

/**
 * Some common types of data you may want to render with ImmutableListView.
 * @see https://facebook.github.io/react-native/docs/listviewdatasource.html#constructor
 */
const data = {
  EMPTY_DATA: Immutable.List(),

  LIST_DATA: Immutable.List(['lists', 'are', 'great']),

  LIST_DATA_NESTED: Immutable.List([['so', 'are'], ['nested', 'lists']]),

  MAP_DATA_LIST_ROWS: Immutable.fromJS({
    first: ['m', 'a', 'p'],
    second: ['foo'],
    third: [],
    fourth: ['bar'],
  }),

  MAP_DATA_MAP_ROWS: Immutable.fromJS({
    first: {
      row1: 'data 1',
      row2: 'data 2',
    },
    second: {},
  }),

  SET_DATA: Immutable.Set(['one', 'two', 'three']),

  RANGE_DATA: Immutable.Range(3, 10, 3),
};

const renderers = {
  /**
   * @param {*} rowData
   */
  renderRow(rowData) {
    return <Text>{JSON.stringify(rowData)}</Text>;
  },

  // eslint-disable-next-line react/prop-types
  renderItem({ item }) {
    return <Text>{JSON.stringify(item)}</Text>;
  },

  /**
   * @param {Immutable.Iterable} sectionData
   * @param {String} category
   */
  renderSectionHeader(sectionData, category) {
    return <Text header>{`${category} (${sectionData.size} items)`}</Text>;
  },
};

const mocks = {
  /**
   * Mock ScrollView so that it doesn't contain any props when rendered by ListView.
   * This is useful for comparison between ListView and ImmutableListView.
   *
   * @returns {ImmutableListView}
   */
  getImmutableListViewWithoutProps() {
    jest.resetModules();

    // eslint-disable-next-line react/prop-types
    const mockScrollView = props =>
      React.createElement('ScrollView', {}, props.children);
    jest.doMock('ScrollView', () => mockScrollView);

    // eslint-disable-next-line global-require
    return require('./ImmutableListView').default;
  },
};

const expectors = {
  expectToMatchSnapshotWithData(immutableData, shouldRenderSectionHeaders) {
    const renderSectionHeaderProps = shouldRenderSectionHeaders
      ? { renderSectionHeader: renderers.renderSectionHeader }
      : {};

    const tree = renderer
      .create(
        <ImmutableListView
          immutableData={immutableData}
          renderRow={renderers.renderRow}
          {...renderSectionHeaderProps}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  },

  expectVirtualizedToMatchSnapshotWithData(immutableData) {
    const tree = renderer
      .create(
        <ImmutableVirtualizedList
          immutableData={immutableData}
          renderItem={renderers.renderItem}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  },

  expectToMatchListViewWithData(immutableData, shouldRenderSectionHeaders) {
    const MockedImmutableListView = mocks.getImmutableListViewWithoutProps();

    const dataSource = new ListView.DataSource({
      rowHasChanged: (r1, r2) => r1 !== r2,
      sectionHeaderHasChanged: (s1, s2) => s1 !== s2,
    });

    const renderSectionHeaderProps = shouldRenderSectionHeaders
      ? { renderSectionHeader: renderers.renderSectionHeader }
      : {};

    const immutableTree = renderer
      .create(
        <MockedImmutableListView
          immutableData={immutableData}
          renderRow={renderers.renderRow}
          renderEmptyInList={null}
          {...renderSectionHeaderProps}
        />
      )
      .toJSON();

    const updatedDataSource = dataSource.cloneWithRows(immutableData.toJS());
    const regularTree = renderer
      .create(
        <ListView
          dataSource={updatedDataSource}
          renderRow={renderers.renderRow}
          {...renderSectionHeaderProps}
        />
      )
      .toJSON();

    expect(immutableTree).toEqual(regularTree);
  },
};

export { data, renderers, mocks, expectors };
