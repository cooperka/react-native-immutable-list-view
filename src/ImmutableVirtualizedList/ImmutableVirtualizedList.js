import Immutable from 'immutable';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import { Text, VirtualizedList } from 'react-native';

import styles from '../styles';
import utils from '../utils';

import { EmptyVirtualizedList } from './EmptyVirtualizedList';

/**
 * A VirtualizedList capable of displaying {@link https://facebook.github.io/immutable-js/ Immutable} data
 * out of the box.
 */
// eslint-disable-next-line react/prefer-stateless-function
class ImmutableVirtualizedList extends PureComponent {

  static propTypes = {
    // Pass through any props that VirtualizedList would normally take.
    ...VirtualizedList.propTypes,

    /**
     * The immutable data to be rendered in a VirtualizedList.
     */
    // eslint-disable-next-line consistent-return
    immutableData: (props, propName, componentName) => {
      // Note: It's not enough to simply validate PropTypes.instanceOf(Immutable.Iterable),
      // because different imports of Immutable.js across files have different class prototypes.
      // TODO: Add support for Immutable.Map, etc.
      if (Immutable.Map.isMap(props[propName])) {
        return new Error(`Invalid prop ${propName} supplied to ${componentName}: Support for Immutable.Map is coming soon. For now, try an Immutable List, Set, or Range.`);
      } else if (!utils.isImmutableIterable(props[propName])) {
        return new Error(`Invalid prop ${propName} supplied to ${componentName}: Must be instance of Immutable.Iterable.`);
      }
    },

    /**
     * A plain string, or a function that returns some {@link PropTypes.element}
     * to be rendered in place of a `VirtualizedList` when there are no items in the list.
     *
     * Things like pull-refresh functionality will be lost unless explicitly supported by your custom component.
     * Consider `renderEmptyInList` instead if you want this.
     *
     * It will be passed all the original props of the ImmutableVirtualizedList.
     */
    renderEmpty: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),

    /**
     * A plain string, or a function that returns some {@link PropTypes.element}
     * to be rendered inside of an `EmptyVirtualizedList` when there are no items in the list.
     *
     * This allows pull-refresh functionality to be preserved.
     *
     * It will be passed all the original props of the ImmutableVirtualizedList.
     */
    renderEmptyInList: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  };

  static defaultProps = {
    ...VirtualizedList.defaultProps,

    renderEmptyInList: 'No data.',
  };

  getVirtualizedList() {
    return this.virtualizedListRef;
  }

  scrollToEnd = (...args) =>
    this.virtualizedListRef && this.virtualizedListRef.scrollToEnd(...args);

  scrollToIndex = (...args) =>
    this.virtualizedListRef && this.virtualizedListRef.scrollToIndex(...args);

  scrollToItem = (...args) =>
    this.virtualizedListRef && this.virtualizedListRef.scrollToItem(...args);

  scrollToOffset = (...args) =>
    this.virtualizedListRef && this.virtualizedListRef.scrollToOffset(...args);

  recordInteraction = (...args) =>
    this.virtualizedListRef && this.virtualizedListRef.recordInteraction(...args);

  renderEmpty() {
    const {
      immutableData, renderEmpty, renderEmptyInList, contentContainerStyle,
    } = this.props;

    const shouldTryToRenderEmpty = renderEmpty || renderEmptyInList;
    if (shouldTryToRenderEmpty && utils.isEmptyListView(immutableData)) {
      if (renderEmpty) {
        if (typeof renderEmpty === 'string') {
          return <Text style={[styles.emptyText, contentContainerStyle]}>{renderEmpty}</Text>;
        }
        return renderEmpty(this.props);
      }
      if (renderEmptyInList) {
        if (typeof renderEmptyInList === 'string') {
          const { renderItem, ...passThroughProps } = this.props;
          return <EmptyVirtualizedList {...passThroughProps} emptyText={renderEmptyInList} />;
        }
        return <EmptyVirtualizedList {...this.props} renderItem={() => renderEmptyInList(this.props)} />;
      }
    }

    return null;
  }

  render() {
    const { immutableData, renderEmpty, renderEmptyInList, ...passThroughProps } = this.props;

    return this.renderEmpty() || (
      <VirtualizedList
        ref={(component) => { this.virtualizedListRef = component; }}
        data={immutableData}
        getItem={(items, index) => utils.getValueFromKey(index, items)}
        getItemCount={(items) => (items ? items.size || 0 : 0)}
        keyExtractor={(item, index) => String(index)}
        {...passThroughProps}
      />
    );
  }

}

export default ImmutableVirtualizedList;
