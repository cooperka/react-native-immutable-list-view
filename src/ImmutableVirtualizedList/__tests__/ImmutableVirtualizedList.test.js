import React from 'react';
import renderer from 'react-test-renderer';

import { data, renderers, expectors } from '../../test-utils';

import ImmutableVirtualizedList from '../ImmutableVirtualizedList';

describe('ImmutableVirtualizedList', () => {
  it('renders with empty data', () => {
    expectors.expectVirtualizedToMatchSnapshotWithData(data.EMPTY_DATA);
  });

  it('renders basic List', () => {
    expectors.expectVirtualizedToMatchSnapshotWithData(data.LIST_DATA);
  });

  it('renders nested List', () => {
    expectors.expectVirtualizedToMatchSnapshotWithData(data.LIST_DATA_NESTED);
  });

  it('renders basic Range', () => {
    expectors.expectVirtualizedToMatchSnapshotWithData(data.RANGE_DATA);
  });
});

describe('ImmutableVirtualizedList with renderEmpty', () => {
  it('renders normally when there are some items', () => {
    const tree = renderer.create(
      <ImmutableVirtualizedList
        immutableData={data.LIST_DATA}
        renderItem={renderers.renderRow}
        renderEmpty={() => renderers.renderRow('No items')}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders empty with a function', () => {
    const tree = renderer.create(
      <ImmutableVirtualizedList
        immutableData={data.EMPTY_DATA}
        renderItem={renderers.renderRow}
        renderEmpty={() => renderers.renderRow('No items')}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders empty with a string', () => {
    const color = 'red';

    const tree = renderer.create(
      <ImmutableVirtualizedList
        immutableData={data.EMPTY_DATA}
        renderItem={renderers.renderRow}
        renderEmpty="No items"
        contentContainerStyle={{ color }}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('doesn\'t render empty with null', () => {
    const tree = renderer.create(
      <ImmutableVirtualizedList
        immutableData={data.EMPTY_DATA}
        renderItem={renderers.renderRow}
        renderEmpty={null}
        renderEmptyInList={null}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});

describe('ImmutableVirtualizedList with renderEmptyInList', () => {
  it('renders normally when there are some items', () => {
    const tree = renderer.create(
      <ImmutableVirtualizedList
        immutableData={data.LIST_DATA}
        renderItem={renderers.renderRow}
        renderEmptyInList={() => renderers.renderRow('No items')}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders empty with a function', () => {
    const tree = renderer.create(
      <ImmutableVirtualizedList
        immutableData={data.EMPTY_DATA}
        renderItem={renderers.renderRow}
        renderEmptyInList={() => renderers.renderRow('No items')}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders empty with a string', () => {
    const color = 'red';

    const tree = renderer.create(
      <ImmutableVirtualizedList
        immutableData={data.EMPTY_DATA}
        renderItem={renderers.renderRow}
        renderEmptyInList="No items"
        contentContainerStyle={{ color }}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('doesn\'t render empty with null', () => {
    const tree = renderer.create(
      <ImmutableVirtualizedList
        immutableData={data.EMPTY_DATA}
        renderItem={renderers.renderRow}
        renderEmpty={null}
        renderEmptyInList={null}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
