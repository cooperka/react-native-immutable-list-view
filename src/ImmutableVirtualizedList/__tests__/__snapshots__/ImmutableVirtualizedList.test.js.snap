// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImmutableVirtualizedList renders basic List 1`] = `
<RCTScrollView
  data={
    Immutable.List [
      "lists",
      "are",
      "great",
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        "lists"
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        "are"
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        "great"
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ImmutableVirtualizedList renders basic Range 1`] = `
<RCTScrollView
  data={Immutable.Seq [â€¦]}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        3
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        6
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        9
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ImmutableVirtualizedList renders nested List 1`] = `
<RCTScrollView
  data={
    Immutable.List [
      Array [
        "so",
        "are",
      ],
      Array [
        "nested",
        "lists",
      ],
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        ["so","are"]
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        ["nested","lists"]
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ImmutableVirtualizedList renders with empty data 1`] = `
<RCTScrollView
  data={
    Immutable.List [
      Immutable.List [
        undefined,
        "No data.",
        "No data.",
      ],
    ]
  }
  disableVirtualization={false}
  emptyText="No data."
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "padding": 8,
            "textAlign": "center",
          }
        }
      >
        No data.
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ImmutableVirtualizedList with renderEmpty doesn't render empty with null 1`] = `
<RCTScrollView
  data={Immutable.List []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View />
</RCTScrollView>
`;

exports[`ImmutableVirtualizedList with renderEmpty renders empty with a function 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  "No items"
</Text>
`;

exports[`ImmutableVirtualizedList with renderEmpty renders empty with a string 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "padding": 8,
        "textAlign": "center",
      },
      Object {
        "color": "red",
      },
    ]
  }
>
  No items
</Text>
`;

exports[`ImmutableVirtualizedList with renderEmpty renders normally when there are some items 1`] = `
<RCTScrollView
  data={
    Immutable.List [
      "lists",
      "are",
      "great",
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        {"item":"lists","index":0,"separators":{}}
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        {"item":"are","index":1,"separators":{}}
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        {"item":"great","index":2,"separators":{}}
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ImmutableVirtualizedList with renderEmptyInList doesn't render empty with null 1`] = `
<RCTScrollView
  data={Immutable.List []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View />
</RCTScrollView>
`;

exports[`ImmutableVirtualizedList with renderEmptyInList renders empty with a function 1`] = `
<RCTScrollView
  data={
    Immutable.List [
      Immutable.List [
        undefined,
        [Function],
        "No data.",
      ],
    ]
  }
  disableVirtualization={false}
  emptyText="No data."
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        "No items"
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ImmutableVirtualizedList with renderEmptyInList renders empty with a string 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "color": "red",
    }
  }
  data={
    Immutable.List [
      Immutable.List [
        undefined,
        "No items",
        "No items",
      ],
    ]
  }
  disableVirtualization={false}
  emptyText="No items"
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "padding": 8,
            "textAlign": "center",
          }
        }
      >
        No items
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ImmutableVirtualizedList with renderEmptyInList renders normally when there are some items 1`] = `
<RCTScrollView
  data={
    Immutable.List [
      "lists",
      "are",
      "great",
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        {"item":"lists","index":0,"separators":{}}
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        {"item":"are","index":1,"separators":{}}
      </Text>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        {"item":"great","index":2,"separators":{}}
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
